packages/contractkit/src/identity/types.ts

```diff
@@ -0,0 +1,102 @@
+import { findFirst } from 'fp-ts/lib/Array'
+import {
+  chain,
+  Either,
+  fromOption,
+  map,
+  mapLeft,
+  parseJSON,
+  toError,
+  tryCatch,
+} from 'fp-ts/lib/Either'
+import { pipe } from 'fp-ts/lib/pipeable'
+import * as t from 'io-ts'
+
+export enum ClaimTypes {
+  ATTESTATION_SERVICE_URL = 'ATTESTATION_SERVICE_URL',
+  DNS = 'DNS',
+  NAME = 'NAME',
+  PROFILE_PICTURE = 'PROFILE_PICTURE',
+  TWITTER = 'TWITTER',
+}
+
+const UrlType = t.string
+const SignatureType = t.string
+const TimestampType = t.number
+
+const AttestationServiceURLClaimType = t.type({
+  type: t.literal(ClaimTypes.ATTESTATION_SERVICE_URL),
+  timestamp: TimestampType,
+  url: UrlType,
+})
+
+const DnsClaimType = t.type({
+  type: t.literal(ClaimTypes.DNS),
+  timestamp: TimestampType,
+  domain: t.string,
+})
+
+const NameClaimType = t.type({
+  type: t.literal(ClaimTypes.NAME),
+  timestamp: TimestampType,
+  name: t.string,
+})
+
+export const ClaimType = t.union([AttestationServiceURLClaimType, DnsClaimType, NameClaimType])
+export const SignedClaimType = t.type({
+  payload: ClaimType,
+  signature: SignatureType,
+})
+
+export const IdentityMetadataType = t.type({
+  claims: t.array(SignedClaimType),
+})
+
+export type Claim = t.TypeOf<typeof ClaimType>
+export type SignedClaim = t.TypeOf<typeof SignedClaimType>
+export type AttestationServiceURLClaim = t.TypeOf<typeof AttestationServiceURLClaimType>
+export type NameClaim = t.TypeOf<typeof NameClaimType>
+export type IdentityMetadata = t.TypeOf<typeof IdentityMetadataType>
+
+export class ValidationError extends Error {}
+// tslint:disable-next-line: max-classes-per-file
+export class ClaimNotFoundError extends Error {}
+
+export const findClaim = (type: ClaimTypes) => (data: IdentityMetadata) =>
+  pipe(
+    data.claims,
+    findFirst((claim: SignedClaim) => claim.payload.type === type),
+    fromOption(() => new ClaimNotFoundError('Could not find claim')),
+    map((signedClaim) => signedClaim.payload)
+  )
+
+export const validateMetadata = (data: any): Either<ValidationError, IdentityMetadata> => {
+  const result = IdentityMetadataType.decode(data)
+  return mapLeft(toError)(result)
+}
+
+export function serializeMetadata(data: IdentityMetadata) {
+  return JSON.stringify({
+    claims: data.claims.map((claim) => ({
+      payload: JSON.stringify(claim.payload),
+      signature: claim.signature,
+    })),
+  })
+}
+
+export const parseMetadata = (str: string) => {
+  const firstPass = parseJSON(str, toError)
+  return chain(deserializePayloads)(firstPass)
```

Comment:
it seems to me that we are parsingJSON, and then we assume it has some schema.

In particual, we assume `data.claims`

Isn't io-ts supposed to do the validation for use?

Reply:
`io-ts` can validate a JS object to be of a particular shape, but unless you produce specific types that parse from a string, `io-ts` does not by default do the JSON parsing itself. So in practice, we would generally do `parseMetadata` in conjunction with `validateMetadata`

Reply:
Yes, but you are iterating it (on deserializePayloads), asssuming that (a) is an object, (b) has a `claims` property which is an array. 

Validating that should be what the validator library does (either io-ts or any other json validator library)

Reply:
Oh yeah, you are right. Right now this code just assumes an object that has the claims property. It is because the payload itself is serialized twice, which does not fit well into the built-in types (plus eventually verifying the signatures). It seems to me that instead of shoe-horning it into the types, I would actually keep this separate, what do you think?